hazelcast/src/main/java/com/hazelcast/internal/util/collection/ImmutableLazyEntry.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright (c) 2008-2023, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.internal.util.collection;
+
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.core.HazelcastInstanceAware;
+import com.hazelcast.internal.nio.IOUtil;
+import com.hazelcast.internal.serialization.Data;
+import com.hazelcast.internal.serialization.SerializationService;
+import com.hazelcast.nio.ObjectDataInput;
+import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
+import com.hazelcast.spi.impl.SerializationServiceSupport;
+
+import java.io.IOException;
+import java.util.Map.Entry;
+import java.util.Objects;
+
+/**
+ * Implementation of {@link Entry} which lazily deserializes keyData and valueData on access.
+ * This class is not thread-safe.
```

Comment:
what is the problem with being thread-safe? Just curious

Reply:
1. Concurrent calls to `getKey`/`Value` can deserialize the key/value multiple times (not a huge issue, just doing more work than needed)
2. The constructor doesn't provide safe publication, so passing the newly created instance from one thread to another needs to ensure HB relationship somehow, otherwise you might see the keyData/valueData fields as null.